integration_test.go

```diff
@@ -0,0 +1,279 @@
+	cmd := exec.Command(os.Args[0], "-test.run=TestIntegrationServer")
+
+	// Set the feature toggle for the integration test server
+	cmd.Env = []string{intSrvFeatureToggle + "=1"}
+
+	return cmd
+}
+
+func TestIntegrationRestart(t *testing.T) {
+	intServer := createIntegrationServerTest()
+	msgs := make(chan string)
+	// Start test integration server
+	require.NoError(t, startIntegrationServer(intServer, msgs), "starting integration test server")
+	ppid := intServer.Process.Pid
+	defer teardownServer(ppid)
+	go intServer.Wait()
+
+	// The pid is printed out from the integation test server once it has fully started up
+	// Check for this message so we can start sending requests/signals to the server
+	msg, err := waitForMessage(msgs, 3*time.Second)
+	require.NoError(t, err, "wait for server startup")
+	msgPid, _ := strconv.Atoi(msg) // don't care about error here as we're just going to compare msgPid to the parentpid anyway
+	require.Equal(t, ppid, msgPid, "Unexpected message down messages channel")
+
+	// Wait for the server to actually come up and respond to requests before we send any important ones
+	require.NoError(t, waitForServer(fmt.Sprintf("http://localhost:%s%s", intSrvPort, intSrvAlivePath)), "wait for server to accept requests")
+
+	// Send a long running request to the test server
+	lReqWait := &sync.WaitGroup{}
+	lReqWait.Add(1)
+	var (
+		lResp   *http.Response
+		lReqErr error
+	)
+	go func() {
+		lResp, lReqErr = sendIntegrationTestRequest(2 * time.Second)
+		lReqWait.Done()
+	}()
+
+	msg, err = waitForMessage(msgs, 1*time.Second)
+	require.NoError(t, err, "wait for long request to arrive")
+	require.Equal(t, intSrvReqRecv, msg, "unexpected message while waiting for long request to arrive")
+
+	// Send a signal to the test server to restart
+	require.NoError(t, intServer.Process.Signal(syscall.SIGHUP), "send signal to restart")
+
+	// The child server should also print it's pid out once it's started up, so look for this
+	msg, err = waitForMessage(msgs, 3*time.Second)
+	require.NoError(t, err, "wait for child server startup")
+	cpid, _ := strconv.Atoi(msg) // don't care about error here as we're just going to check msgPid isn't zero
+	require.NotZero(t, msgPid, "Unexpected message down messages channel")
+	defer teardownServer(cpid) // set up the child server to be torn down on test exit
+
+	// Send a quick request to the server. The child server should now be up and running, accepting requests,
+	// so check that it is the child PID that appears in the response and not the parent
+	qResp, qReqErr := sendIntegrationTestRequest(0 * time.Second)
+	require.NoError(t, qReqErr, "sending quick request to child server")
+	require.Equal(t, http.StatusOK, qResp.StatusCode, "response code from long request")
+
+	qBody, err := ioutil.ReadAll(qResp.Body)
+	qResp.Body.Close()
+	require.NoError(t, err, "read from quick request body")
+	require.NotContains(t, string(qBody), fmt.Sprintf("%d", ppid), "quick response contained wrong PID")
```

Comment:
In general don't use %d for int's use %v but also strconv.Atoi(ppid) would be better here.

Reply:
guessing you meant itoa here